### ISSUE TITLE
Azure Files Share create fails with 403

### ISSUE THREAD

**Issue by doherty100** [hashicorp/terraform-provider-azurerm#25527]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.7.5

### AzureRM Provider Version

3.98.0 or 3.99.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_share

### Terraform Configuration Files

```hcl
resource "azurerm_storage_share" "storage_share_01" {
  name                 = var.storage_share_name
  storage_account_name = var.storage_account_name
  quota                = var.storage_share_quota_gb
}
```

### Debug Output/Panic Output

```shell
╷
│ Error: checking for existing File Share "myfileshare" (Account "Account \"staymunn1rch5rh\" (IsEdgeZone false / ZoneName \"\" / Subdomain Type \"file\" / DomainSuffix \"core.windows.net\")"): executing request: unexpected status 403 (403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.) with response: <?xml version="1.0" encoding="utf-8"?><Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
│ RequestId:99d9f348-c01a-0034-0f1e-88fb57000000
│ Time:2024-04-06T12:32:18.4515117Z</Message><AuthenticationErrorDetail>The MAC signature found in the HTTP request 'Mx+0gs4iAFYD4R4Xoy82e983g53R6Fx0qnoYKbuZXqw=' is not the same as any computed signature. Server used following string to sign: 'GET
│ 
│ 
│ 
│ 
│ application/xml; charset=utf-8
│ 
│ 
│ 
│ 
│ 
│ 
│ x-ms-date:Sat, 06 Apr 2024 12:32:18 GMT
│ x-ms-version:2023-11-03
│ /staymunn1rch5rh/myfileshare
│ restype:share'.</AuthenticationErrorDetail></Error>
│ 
│   with azurerm_storage_share.storage_share_01,
│   on 050-storage-share.tf line 2, in resource "azurerm_storage_share" "storage_share_01":
│    2: resource "azurerm_storage_share" "storage_share_01" {
│ 
╵
```

### Expected Behaviour

Resource should provision successfully.

### Actual Behaviour

╷
│ Error: checking for existing File Share "myfileshare" (Account "Account \"staymunn1rch5rh\" (IsEdgeZone false / ZoneName \"\" / Subdomain Type \"file\" / DomainSuffix \"core.windows.net\")"): executing request: unexpected status 403 (403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.) with response: <?xml version="1.0" encoding="utf-8"?><Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
│ RequestId:99d9f348-c01a-0034-0f1e-88fb57000000
│ Time:2024-04-06T12:32:18.4515117Z</Message><AuthenticationErrorDetail>The MAC signature found in the HTTP request 'Mx+0gs4iAFYD4R4Xoy82e983g53R6Fx0qnoYKbuZXqw=' is not the same as any computed signature. Server used following string to sign: 'GET
│ 
│ 
│ 
│ 
│ application/xml; charset=utf-8
│ 
│ 
│ 
│ 
│ 
│ 
│ x-ms-date:Sat, 06 Apr 2024 12:32:18 GMT
│ x-ms-version:2023-11-03
│ /staymunn1rch5rh/myfileshare
│ restype:share'.</AuthenticationErrorDetail></Error>
│ 
│   with azurerm_storage_share.storage_share_01,
│   on 050-storage-share.tf line 2, in resource "azurerm_storage_share" "storage_share_01":
│    2: resource "azurerm_storage_share" "storage_share_01" {
│ 
╵

### Steps to Reproduce

Upgrade to azurerm v3.98.0 then attempt to provision an Azure files Share.

### Important Factoids

Breaking change with azurerm v3.98.0, works 3.97.1

**Comment by doherty100** [comment]
Upgrading to azurerm v3.99.0 did not resolve the issue.

**Comment by mifitous** [comment]
> Upgrading to azurerm v3.99.0 did not resolve the issue.

same here

**Comment by manicminer** [comment]
I've been trying to reproduce this bug without success so far. In v3.98.0, the Resource Manager SDK for storage accounts was changed to use [hashicorp/go-azure-sdk](https://github.com/hashicorp/go-azure-sdk) - this should have had no user facing effects but is the only related change that I can find which to which this issue may be attributable.

Whilst stepping through the handling of storage authentication, I noticed a small discrepancy from our previous behavior, in that since v3.98.0, the storage account key used for authenticating to the Storage API is now the second account key, where previously it was always the first key. I don't fully understand why this would cause issues, but I have opened a PR to revert this behavior at https://github.com/hashicorp/terraform-provider-azurerm/pull/25652

For anyone affected by this issue: would you be able to build and test that PR branch (`bugfix/storage-keyauth-403`) and advise whether the issue remains, or whether it goes away for you? In the meantime I will be working on some additional logging to help track this down in the wild. Thanks!

**Comment by DuncanCasteleyn** [comment]
We have the same issue with the data provider to read a share on the plan step. everything works as long as we don't upgrade to version 3.98.0 or 3.99.0:

values have been replaced by redacted.

Example
```terraform
data "azurerm_storage_share" "redacted" {
  provider             = azurerm.redacted
  name                 = "redacted"
  storage_account_name = "redacted"
}
```

```
╷
│ Error: retrieving https://redacted.file.core.windows.net/redacted: executing request: unexpected status 403 (403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.) with response: ﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
│ RequestId:redacted
│ Time:2024-04-18Tredacted</Message><AuthenticationErrorDetail>The MAC signature found in the HTTP request 'redacted' is not the same as any computed signature. Server used following string to sign: 'GET
│ 
│ 
│ 
│ 
│ application/xml; charset=utf-8
│ 
│ 
│ 
│ 
│ 
│ 
│ x-ms-date:Thu, 18 Apr 2024 redacted
│ x-ms-version:2023-11-03
│ /redacted/redacted
│ restype:share'.</AuthenticationErrorDetail></Error>
│ 
│   with data.azurerm_storage_share.redacted,
│   on data.tf line 55, in data "azurerm_storage_share" "redacted":
│   55: data "redacted" "redacted" {
│ 
╵
```

**Comment by doherty100** [comment]
@manicminer this issue has been resolved in azurerm v3.100.0.

**Comment by DuncanCasteleyn** [comment]
I can also confirm that this issue is resolved in v3.100.0.

**Comment by manicminer** [comment]
@doherty100, @DuncanCasteleyn Great! Thanks for reporting back, and I'm glad this has resolved the 403 for you. Appreciate the feedback.
